### ISSUE TITLE
adding alloc_id tag to sidecar `envoy_stats_tags`

### ISSUE THREAD

**Issue by fredwangwang** [hashicorp/nomad#12326]
### Proposal

Would like to have alloc_id tag (like other nomad allocation level metrics) on the metrics emitted by the mesh sidecars.

### Use-cases

Atm all the tags emitted by mesh sidecar (envoy) only have consul service info, which does not differentiate each individual allocations.
These metrics are good for getting an overall aggregated view of the traffic patterns, but not so useful for inspecting traffic for individual service allocations.
Adding `alloc_id` to the metrics emitted by mesh sidecar would enable that.

**Comment by idrennanvmware** [comment]
Thanks Huan - I might suggest that we expand this request to something more generic.

Overall it would be great to be able to inject static tags in to mesh side cars so that they're emitted as point tags to telemetry systems. Alloc_id is one thing but I can see use cases for other things too.  This ability gives timeseries systems some really nice query capabilities.

**Comment by tgross** [comment]
Hi @fredwangwang and @idrennanvmware! The sidecars are just tasks that we add to the job at the time they're submitted (see [`job_endpoint_hook_connect.go`](https://github.com/hashicorp/nomad/blob/main/nomad/job_endpoint_hook_connect.go)). So they should emit the same telemetry as other tasks. Can y'all clarify a bit as to which metrics you're looking for on sidecars that aren't already tagged the same way all other tasks are? 

As an aside, the [`sidecar_task.config`](https://www.nomadproject.io/docs/job-specification/sidecar_task#config) already provides a generic facility for making changes like labelling and metadata for sidecar tasks, but this issue seems to be specifically about Nomad-generated metrics, right?

**Comment by tgross** [comment]
One of my colleagues has pointed out that [`connect.sidecar_service.proxy.config`](https://www.nomadproject.io/docs/job-specification/proxy#config) may give you the access to the `envoy_stats_tags`, but this is opaque to Nomad. I'm not sure we have a ready-to-use configuration option for you without some more testing, but that might be something you'd like to experiment with if you have time to try it out.

**Comment by fredwangwang** [comment]
Just tried (not fully working)

Btw thanks for pointing out `envoy_stats_tags` totally forgot about that :)

and here is my updated nomad config:
```

            config {
              envoy_stats_tags = ["alloc_id=${NOMAD_ALLOC_DIR}"]
            }
```

here is the rendered envoy bootstrap config:

```
    "stats_tags": [
      {
        "tag_name": "alloc_id",
        "fixed_value": "${NOMAD_ALLOC_ID}"
      },
```

soo the nomad variables does not gets interpolated, most likely as you said it is currently opaque to nomad.

**Comment by tgross** [comment]
Ok, thanks. That should be enough for us to go on in terms of making this a roadmap item. (cc @mikenomitch as I know he had some interest in this one.)

**Comment by fredwangwang** [comment]
thanks @tgross! Also, would it makes sense to always add `alloc_id` to the envoy metrics without user making the addition (even with the interpolation available)? 

As said our use case is using the envoy metrics to correlate to the allocation performance, which I would think a rather common thing to do with these metrics.

**Comment by tgross** [comment]
That seems sensible, as it's a bit of a pain to configure by hand for sure.

**Comment by fredwangwang** [comment]
cool! we can try put up a pr for adding alloc_id by default if you haven't started it yet, let me know

**Comment by tgross** [comment]
Yes, I'd be happy to review a PR for it @fredwangwang!

**Comment by idrennanvmware** [comment]
@tgross  - @gulavanir and I took a crack at this (thanks @fredwangwang for some extra eyes). We aren't Golang experts by any means so please let us know if we need to address anything

**Comment by shoenig** [comment]
Closed by  #12543
